Placement in/box refinement

Stebalien opened the pull request:
Supersedes #1401

[rendered](https://github.com/Stebalien/rfcs/blob/placement-traits/text/0000-placement-traits.md)

eddyb commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,312 @@
+[unresolved]: #unresolved-questions
+
+Does `Place::pointer` need to return an `&mut T` (or some other special pointer)
+to ensure RVO? Should it return `std::ptr::Unique`?
```
I almost feel like `Unique<T, A>` should be `Place<T, A>`'s "filled" counterpart, if both were structs and `A` described how the allocation should be handled.

eddyb commented on text/0000-placement-traits.md:
````diff
@@ -0,0 +1,312 @@
+```
+
+This means that `let boxed_thing: Type<_> = HEAP <- thing` works out of the box
+as long as `HEAP` is an `Allocator` and `Type` implements `Boxer`.
````
Why are we still using `HEAP` as an example?
Isn't there a name like `DefaultAllocator` in one of the allocator RFCs we can pick up and move on?

Stebalien replied:
#39 (closed). It's not mentioned anywhere else.

nagisa replied:
To my knowledge HEAP et al are supposed to be removed ASAP (or never, ever stabilised), therefore this example is not good.

Also, to my knowledge, there should not be things like `HEAP` in the standard library at all, thus making all this `Allocator` thing redundant somewhat?

All in all, to me it seems like this whole section is a bit hasty and we should at least wait for `Allocator` design to get landed first (which doesn’t seem to be finishing anytime soon).

Stebalien replied:
The original intention of `PLACE <- thing` was to allow placing into custom allocators. I'm probably going to change this to https://github.com/rust-lang/rfcs/pull/1426#issuecomment-167160921 which can be added after the allocator API is merged (I think).

eddyb commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,312 @@
+/// unsafe {
+///     std::ptr::write(raw_place, value);
+///     Place::finalize(place)
+/// }
```
For inference to work, all of this (and the `box` desugaring below) should be nested expressions (no blocks and statements), either with some safe `write_and_finalize` function (harder to enforce optimizations like we have on `std::ptr::write`) or with a special HIR node that is lowered to the appropriate MIR/LLVM IR.

eddyb commented:
To expand on the special HIR node idea:
```rust
PLACE <- EXPR // gets desugared into:
ExprEmplace(Placer, PLACE, EXPR)
```
```rust
box EXPR // gets desugared into:
ExprEmplace(Boxer, ExprCall(Default::default, []), EXPR)
```

This way, we could guarantee that 0 unnecessary copies are being made, DST support can be added without having to come up with impossible source desugarings and, probably the best part, we can employ a simpler form of rust-lang/rust#27292 only for `ExprEmplace`:

If we have an expected type, i.e. `ExprEmplace(Placer, _, E): P<A, B, C>`, we check that `P<A, B, C>: Placer<T>`.
In case we do get a match, `T: Sized` must hold and we can use `T` as the expected type for `E`.

Otherwise, it's either a legitimate error, or our `P<A, B, C>` can only be produced by unsizing the result of `ExprEmplace` (e.g. `Box<Trait>`). We can assume it's not an error and attempt to make it compile.

Thankfully, the unsizing coercions relevant here only involve structs (and `Box`, which we can treat as a struct).
This simplifies our job greatly and if more types support unsizing coercions in the future, we can simply choose not to handle them here.

We can take the most conservative stance and replace all type params in `P<A, B, C>` with inference variables to get a condidate for our missing intermediary smart pointer, e.g. `P<X, Y, Z>.

Then, we require that `P<X, Y, Z>: CoerceUnsized<P<A, B, C>> + Placer<T>`.
This will link up all the type params and allow us to type-check the emplace expression.

 E.g. for `Box<Trait, A>`, we require `Box<X, Y>: CoerceUnsized<Box<Trait, A>> + Boxer<T>` which will result in `X = T` and `Y = A` and `T: Trait` obligation.
In the end we get our intermediary smart pointer, `Box<T, A>`, with a `Boxer` impl we can get the associated `Place` from, and a coercion to `Box<Trait, A>`.

Unlike rust-lang/rust#27292, this also works if you replace `Trait` with `[T]`, and does not affect coercions in existing stable code.

mahkoh commented:
Why would the Allocator be constructed in the desugaring and the Boxer have an Allocator parameter? One can simply

```rust
impl<T, A> Boxer<T> for Box<T, A> where A: Allocator+Default
```

Having the Boxer trait know anything about allocators is completely unnecessary and a huge restriction.

mahkoh commented:
@eddyb Doesn't move_val_init already do what you want? Except that move_val_init doesn't unconditionally get translated to a memcpy for some reason.

eddyb commented:
@mahkoh The problem is that you can't desugar to `move_val_init(ptr, EXPR)` because it requires `unsafe` (and you don't want to allow unsafe code in `EXPR`) and it has to be its own statement in a nested block so coercion propagation is pretty much screwed.

Stebalien commented:
@mahkoh, how is this a restriction? `Boxer

mahkoh commented:
The Boxer implementation can acquire an allocator (if it even needs one) by means other than `Default`.

Stebalien commented:
@mahkoh, sorry, hit enter too soon. See my updated post.

mahkoh commented:
@Stebalien All of this can be done without the restrictions imposed by having Boxer know about allocators. E.g.

````rust
let a: Box<_, A1> = box val; // where A1: Default

let a = Box::with_pool(CustomAllocator::new(options)) <- thing;
```

mahkoh commented:
The first line is equivalent to

```rust
let a = Box::with_pool(A1::default()) <- val;
```

Stebalien commented:
@mahkoh, which means that one has to implement a new `Placer`. Regardless, I still don't see how this is a *restriction*. What is the use case for using `box value` without an `Allocator`? Note, the proposed system still supports your syntax regardless.

mahkoh commented:
>Note, the proposed system still supports your syntax regardless.

It doesn't support `box` without a `Default` allocator.

>What is the use case for using box value without an Allocator?

The object might want to use a globally shared allocator stored in a static variable or might want to construct an allocator with custom options.

mahkoh commented:
>which means that one has to implement a new Placer. 

And where is the problem with that?

mahkoh commented:
You also don't need a "new" Placer. See the following pseudocode

```rust
struct Box<T: ?Sized, A>
    where A: Allocator,
{
    alloc: A,
    val: *mut T,
}

struct BoxBuf<T, A>
    where A: Allocator,
{
    alloc: A,
    val: *mut T,
}

impl<T, A> Boxer<T> for Box<T, A>
    where A: Allocator+Default,
{
    type Place = BoxBuf<T, A>
    fn make_place() -> BoxBuf<T, A> {
        let mut alloc = A::default();
        /* ... */
        BoxBuf {
            alloc: alloc,
            val: val,
        }
    }
}

impl<T, A> Placer<T, Box<T, A>> for BoxBuf<T, A>
    where A: Allocator,
{
    type Place = Self;
    fn make_place(self) -> Self { self }
}

impl<T, A> Place<T> for BoxBuf<T, A>
    where A: Allocator,
{
    type Owner = Box<T, A>;
    fn pointer(&mut self) -> *mut T { self.val }
    unsafe fn finalize(self) -> Box<T, A> {
        Box {
            self.alloc,
            self.val,
        }
    }
}
```

It seems that what you actually want is a default implementation of `Placer` for all `Place` implementations.

mahkoh commented:
To be quite honest, I don't think there is a need for the Placer trait at all. The only use it has is that it can be implemented for Allocators so that you can use the `HEAP <- val` syntax to construct arbitrary containers.

This can also be done with the Place trait alone:

```rust
Box::with_pool(HEAP) <- val
```

So the only advantage seems to be the shorter syntax. The most common cases are

* the `box val` case if there is no need for error handling
* the `try!(xyz) <- val` case if there is a need for error handling

Neither of these cases are served by the Placer trait. There is also the problem with that Allocators have no idea how to construct containers and that, the Placer implementation for Allocators is limited by this. This inversion then leaks into the placement traits by having them suddenly depend on Allocators.

Apart from this forced dependency, allocators and placement traits are orthogonal.

eddyb commented:
@mahkoh also consider `vec <- x` and `hash_map.entry(k) <- v`.

mahkoh commented:
@eddyb The first one doesn't work because Placer consumes self. I'll think a bit about the second one.

eddyb commented:
@mahkoh it would work if `vec: &mut Vec<T>` and/or `<-` does autoref on the LHS.

mahkoh commented:
@eddyb Is there a reason that `Place::pointer` cannot do the heavy lifting? (Destroying the current entry in the hashmap if it exists, reserving a slot in the vector.)

Stebalien commented:
@eddyb, note, as of this RFC, it has to be `vec.back() <- x` (unless we get auto referencing).

@mahkoh, FYI, it's `PLACER <- thing;`, not `PLACE <- thing;`.

Re your previous comments, this really just comes down to the common. With the system proposed in the RFC, one would write:

```rust
let boxed: Box<_> = HEAP <- value;
let object = Object::from_object_pool() <- value;
```

With your system, one would write:

```rust
let boxed: Box<_> = Box::with_allocator(HEAP) <- value;
let object: Object = box value;
```

I'll try to come up with a way to make both cases simpler.

eddyb commented:
@mahkoh The primary concern @pnkfelix had, IIRC, is handling panics.

mahkoh commented:
@eddyb I see. Well, I'm not bothered much by the Placer trait itself. It's the idea to implement it for Allocators that seems wrong.

Stebalien commented:
> @eddyb Is there a reason that Place::pointer cannot do the heavy lifting? (Destroying the current entry in the hashmap if it exists, reserving a slot in the vector.)

Also, `Place::pointer` takes self by reference so you'd have to handle the case of user code calling pointer more or less than once.

eddyb commented:
@Stebalien Well that could be made unsafe and the compiler implementation of emplace can do the right thing (TM).

@mahkoh Another "nice" example: `arena <- value` where `arena: &Arena`.

mahkoh commented:
@Stebalien 

>Also, Place::pointer takes self by reference so you'd have to handle the case of user code calling pointer more or less than once.

Just go full `unsafe` in all signatures. Does anyone expect that these traits will be used much outside of the expansions and possibly unsafe code?

>FYI, it's PLACER <- thing;, not PLACE <- thing;.

I'm aware.

>With your system, one would write:

In the first case you don't have to write the type on the lhs because `Box::with_allocator` already contains all the type information:

```rust
let boxed = Box::with_allocator(HEAP) <- value;
```

Stebalien commented:
Sorry, I assumed that we still had a `Placer` to call an unsafe `make_place` on. Assuming we don't have a `Placer` trait and `Vec::back()` returns an `Place`, what happens when the user drops the `Place` without emplacing (`drop(vec.back())`, `back()` is not unsafe)? All `Place`s would have to keep track of whether or not they have allocated or allocate on construction (not necessarily a bad thing...).

---

As for depending on Allocator, I think I see your point. Having a generic operation (box) depend on allocation is unfortunate. However, I do like `HEAP <- stuff` so how about adding an `Allocate` trait? If you implement `Allocate`, you get `allocator <- thing` for all allocators and `box thing` for all allocators implementing `Default`. Alternatively, you can just implement `Boxer` if you don't want to take an allocator.

```rust
trait Allocator {/* Stuff */}

pub unsafe trait Place<Data: ?Sized> {
    type Owner;

    fn pointer(&mut self) -> *mut Data;

    unsafe fn finalize(self) -> Self::Owner;
}

/// Implement on all types that can be allocated.
///
/// This gives you both `let boxed: Self = ALLOC <- thing` and `let boxed: Self = box thing` (for
/// all allocators implementing `Default`.
trait Allocate<Data: ?Sized>: Sized {
    type Place: Place<Data, Owner=Self>;
    type Allocator: Allocator;

    fn allocate(with: Self::Allocator) -> Self::Place
        where Data: Sized;
}

trait Placer<Data: ?Sized, Owner> {
    type Place: Place<Data, Owner=Owner>;
    fn make_place(self) -> Self::Place
        where Data: Sized;
}

trait Boxer<Data: ?Sized>: Sized {
    type Place: Place<Data, Owner=Self>;

    fn make_place() -> Self::Place
        where Data: Sized;
}

impl<T, D, A> Placer<D, A> for T
    where T: Allocator,
          A: Allocate<D, Allocator=T>
{
    type Place = A::Place;

    fn make_place(self) -> Self::Place
        where Self: Sized
    {
        A::allocate(self)
    }
}

impl<A, D> Boxer<D> for A
    where A: Allocate<D>,
          A::Allocator: Default
{
    type Place = A::Place;

    fn make_place() -> Self::Place
        where Self: Sized
    {
        A::allocate(Default::default())
    }
}
```

pnkfelix commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,312 @@
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Does `Place::pointer` need to return an `&mut T` (or some other special pointer)
```
I'm pretty sure it would be a mistake to have it return `&mut T`, since you'd have to add all the caveats like "there isn't *actually* a `T` behind the reference, so don't call methods on it ... or swap it out with another `T` ..."

mahkoh commented:
Why is `Data` not an associated type of `Place`? The traits could be a lot simpler if it were:

```rust
pub unsafe trait Place {
    type Owner;
    type Data: ?Sized;

    fn pointer(&mut self) -> *mut Self::Data;
    unsafe fn finalize(self) -> Self::Owner;
}

pub trait Placer<P>
    where P: Place,
{
    fn make_place(self) -> P;
}

pub trait Boxer: Sized {
    type Place: Place<Owner = Self>;

    fn make_place() -> Self::Place;
}
```

Stebalien commented:
@mahkoh, good point about place. However, `trait Placer<P: Place>` will cause inference problems because it would allow multiple implementations of `Placer` with the same input/output (with different  intermediate `Place`s). Also, `Boxer` without a type parameter wouldn't allow things like:

```rust
impl<T> Boxer for OpaqueBox { // T isn't used
    type Place = MyPlace<T>
    fn make_place() -> Self::Place { /* */ }
}
```

Stebalien commented:
@mahkoh, what do you think about my proposed `Allocate` trait?

nagisa commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,315 @@
+
+## Allocators
+
+Boxer's now make Places from Allocators. This means that any type implementing
```
s/Boxer’s/Boxers/?

nagisa commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,315 @@
+///
+/// If evaluating EXPR fails, then the destructor for the
+/// implementation of Place is run to clean up any intermediate state
+/// (e.g. deallocate box storage, pop a stack, etc).
```
pop a stack *frame*

Stebalien replied:
Fixed.

nagisa commented:
All in all, everything related to `Allocator` looks hasty (since we do not have any design related to them landed yet) but the changes to the placement traits look okay overall.

Stebalien commented:
@nagisa 
> All in all, everything related to Allocator looks hasty (since we do not have any design related to them landed yet) but the changes to the placement traits look okay overall.

Replying again here as my comment was swallowed in the push. IIRC, the original motivation of `in (PLACE) box EXPR` (or the new `PLACE <- EXPR`) was placing in a custom allocator so making sure it's compatible with that use case is important. Note: This RFC makes no assumptions about the definition of the `Allocator` trait, just that one exists. Regardless, I'm going to update the RFC to use the traits I proposed in https://github.com/rust-lang/rfcs/pull/1426#issuecomment-167160921 which means that the allocator integration can be implemented later.

Stebalien commented:
All (specifically, @nagisa),

I've updated the proposal to move the allocator related logic out of the `Boxer` into the `Allocate` trait and addressed the need for a `Placer` trait.

pnkfelix commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,362 @@
+    unsafe fn finalize(self) -> Self::Owner;
+}
+
+/// All types implemeinting this trait can be constructed using using the `box`
```
typo: "implemeinting" --> "implementing"

pnkfelix commented on text/0000-placement-traits.md:
```diff
@@ -0,0 +1,362 @@
+guaranteed to be allocated before the evaluation of the expression on the right
+hand side. DST placement needs this to be able to compute the size of the DST
+before allocating the place. This means that, in the following cases, whether or
+not the `Box` is ever allocated is explicitly undefined:
```
hmm i guess i don't have any way to improve upon this, but I find it worrisome.

Stebalien replied:
The place-expression is still evaluated first. That is, given, `PLACER <- THING`: `PLACER` will always be evaluated before `THING`. This just means that the compiler gets to choose when `Placer::make_place(PLACER)` is evaluated.

I don't like it but don't see any alternative that supports DST placement.

pnkfelix replied:
hmm that seems like a subtle distinction you are making versus what the text says ... that is, I read the phrase "the place in placement in/new" as talking about the LHS of `P <- V` ... is there some way to rephrase to make this clearer?

Stebalien replied:
Sorry about that. What I mean and what I say are not always the same thing. Does it make more sense now?

nikomatsakis commented:
Can someone summarize the state of this RFC with respect to the comment thread? Does the text include any desired updates? Are important "non-desired" updates summarized in the alternatives section? I'd like to make decisions on older RFCs like this one =)

Ericson2314 commented:
I recall there being good ideas here, but it will be a *lot* easier to evaluate once Allocators are implemented, so I suggest waiting for that.

Stebalien commented:
There are also a few updates I want to make to take advantage of specialization. Should I close this for now and re-open it when allocators land (or are you planning on stabilizing placement new before the allocation API)?

pnkfelix commented:
Yeah I haven't moved on landing allocators because I was waiting until the pieces were in place to support them on the library types (see [RFC 1327](https://github.com/rust-lang/rfcs/blob/master/text/1327-dropck-param-eyepatch.md))

But now the most fundamental language change for that has landed, so I hope to make forward progress there soon.

I don't think you need to close this RFC in the meantime.

Ericson2314 commented:
Yay!

ubsan commented:
@nrc @aturon @pnkfelix 

This would be a good RFC for postponement, predicated on getting allocators.

aturon commented:
@ubsan Agreed -- let's revisit this down the line, when more of the substrate has stabilized.

@rfcbot fcp postpone

aturon commented:
OK, I'm going to close this as postponed. Work on placement has largely stalled; we should probably revisit the whole question at some point.
